src/native/corehost/hostmisc/pal.unix.cpp

```diff
@@ -424,35 +417,63 @@ bool pal::get_dotnet_self_registered_dir(pal::string_t* recv)
     FILE* install_location_file = pal::file_open(install_location_file_path, "r");
     if (install_location_file == nullptr)
     {
-        trace::verbose(_X("The install_location file failed to open."));
+        trace::error(_X("The install_location file ['%s'] failed to open: %s."), install_location_file_path.c_str(), pal::strerror(errno));
         return false;
     }
 
-    bool result = false;
+    pal::string_t install_location, architecture_install_location, file_line;
+    bool is_first_line = true;
 
-    char buf[PATH_MAX];
-    char* install_location = fgets(buf, sizeof(buf), install_location_file);
-    if (install_location != nullptr)
+    while (get_line_from_file(install_location_file, file_line))
     {
-        size_t len = pal::strlen(install_location);
+        size_t arch_sep = file_line.find(_X('='));
+        if (arch_sep == pal::string_t::npos)
+        {
+            if (is_first_line)
+            {
+                install_location = file_line;
+                trace::verbose(_X("Found install location path '%s'."), install_location.c_str());
+            }
+            else
+            {
+                trace::error(_X("Only the first line in '%s' may not have an architecture prefix."), install_location_file_path.c_str());
+            }
+
+            is_first_line = false;
+            continue;
+        }
+
+        pal::string_t arch_prefix = file_line.substr(0, arch_sep);
+        pal::string_t path_to_location = file_line.substr(arch_sep + 1);
 
-        // fgets includes the newline character in the string - so remove it.
-        if (len > 0 && len < PATH_MAX && install_location[len - 1] == '\n')
+        trace::verbose(_X("Found architecture-specific install location path: '%s' ('%s')."), path_to_location.c_str(), arch_prefix.c_str());
+        if (arch_prefix == get_arch())
         {
-            install_location[len - 1] = '\0';
+            architecture_install_location = path_to_location;
+            trace::verbose(_X("Found architecture-specific install location path matching the current OS architecture ('%s'): '%s'."), arch_prefix.c_str(), architecture_install_location.c_str());
+            break;
         }
 
-        trace::verbose(_X("Using install location '%s'."), install_location);
+        is_first_line = false;
+    }
+
+    if (architecture_install_location != "")
```

Comment:
If the config file looks like this on x64 machine:
```console
/default/path
x64=
```

The result will be to use `/default/path`. I'm not sure I like that. The user did specify arch specific path, it's just invalid. Basically I think this should be a very similar behavior to something like:
```console
/default/path
x64=////nonsense///
```

Both cases should fail with error saying that the root path is invalid (doesn't exist).

There's also this case:
```console

x64=/dotnet/x64
```

(the first line is empty). If we're running on `arm64` - it should fail, the only question is if it should fail with the "Did not find any install location" or "Install location '' doesn't exist".
(Note that existing apphost will fail with the latter).